lib/dynamic.js

```diff
@@ -0,0 +1,36 @@
+'use strict'
+
+const { addHook } = require('./util')
+const buildSwagger = require('./swagger')
+const buildOpenapi = require('./openapi')
+
+module.exports = function (fastify, opts, next) {
+  const { routes, Ref } = addHook(fastify)
+
+  opts = Object.assign({}, {
+    exposeRoute: false,
+    hiddenTag: 'X-HIDDEN',
+    stripBasePath: true,
+    openapi: {},
+    swagger: {},
+    transform: null
+  }, opts || {})
+
+  if (opts.exposeRoute === true) {
+    const prefix = opts.routePrefix || '/documentation'
+    fastify.register(require('./routes'), { prefix })
+  }
+
+  const cache = {
+    swaggerObject: null,
+    swaggerString: null
+  }
+
+  if (Object.keys(opts.openapi).length > 0 && opts.openapi.constructor === Object) {
+    fastify.decorate('swagger', buildOpenapi(opts, routes, Ref, cache, next))
```

Comment:
This doesn't look correct. Swagger is very much specifically OpenAPI 2.0, so using this name here seems misleading.

Reply:
Maybe it's done for backward-compatibility, so there will be no need in major version bump.

Reply:
If that's the case, I would leave single `fastify.decorate` line but resolve actual value before calling it.
Don't think it would be breaking either way, as OpenAPI wasn't supported at all in the past.

Reply:
I think it have no difference between assigning the function before or after. It should use either one of them but not both.

For performance, I assume all the work should be done before ```ready``` event is fired. So, the user can call the right function on the ```fastify.ready```.

Reply:
Yeah, I'm not suggesting anything drastic here, just for me this looks somewhat neater:

```
  const decoratorValue = resolveDecoratorValue(/*params*/)
  fastify.decorate('swagger', decoratorValue())
```

Could be a ternary operator, but judging by the length of condition, might get out of hand.